pkg/channel/producer/producer.go

```diff
@@ -0,0 +1,161 @@
+package producer
+
+import (
+	"errors"
+	"github.com/cloudevents/sdk-go/v1/cloudevents"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"knative.dev/eventing-kafka/pkg/channel/health"
+	"knative.dev/eventing-kafka/pkg/channel/message"
+	"knative.dev/eventing-kafka/pkg/channel/util"
+	kafkaproducer "knative.dev/eventing-kafka/pkg/common/kafka/producer"
+	"knative.dev/eventing-kafka/pkg/common/prometheus"
+	"go.uber.org/zap"
+	eventingChannel "knative.dev/eventing/pkg/channel"
+)
+
+// Package Variables
+var (
+	logger         *zap.Logger
+	kafkaProducer  kafkaproducer.ProducerInterface
+	stopChannel    chan struct{}
+	stoppedChannel chan struct{}
+	metrics        *prometheus.MetricsServer
+)
+
+// Wrapper Around Common Kafka Producer Creation To Facilitate Unit Testing
+var createProducerFunctionWrapper = func(brokers string, username string, password string) (kafkaproducer.ProducerInterface, error) {
+	return kafkaproducer.CreateProducer(brokers, username, password)
+}
+
+// Initialize The Producer
+func InitializeProducer(lgr *zap.Logger, brokers string, username string, password string, metricsServer *prometheus.MetricsServer, healthServer *health.Server) error {
+
+	// Set Package Level Logger Reference
+	logger = lgr
+
+	// Mark The Kafka Producer As Ready
+	healthServer.SetProducerReady(false)
+
+	// Create The Producer Using The Specified Kafka Authentication
+	producer, err := createProducerFunctionWrapper(brokers, username, password)
+	if err != nil {
+		logger.Error("Failed To Create Kafka Producer - Exiting", zap.Error(err))
+		return err
+	}
+
+	// Assign The Producer Singleton Instance
+	logger.Info("Successfully Created Kafka Producer")
+	kafkaProducer = producer
+
+	// Reset The Stop Channels
+	stopChannel = make(chan struct{})
+	stoppedChannel = make(chan struct{})
+
+	// Used For Reporting Kafka Metrics
+	metrics = metricsServer
+
+	// Fork A Go Routine To Process Kafka Events Asynchronously
+	logger.Info("Starting Kafka Producer Event Processing")
+	go processProducerEvents(healthServer)
```

Comment:
does this need to have a context for a clean shutdown?

Reply:
We've currently got a couple of shutdown channels managing the lifecycle.  We could look at using contexts but I'd rather do that later on at the larger (full project) level as we continue to align/standardize.  Sound acceptable?